Add EBPMT for PMT pairing in EventBuilding

fengyvoid opened the pull request:
EBPMT Tool in EventBuildingV2 tool set.

This is splitted from PR #307, I agree that it's definitely easier to submit tools one by one and not necessary to submit by a whole tool chain. I will also list the changes based on your comments in #307.

Changes based on your comments:

· Can we translate EBPMT.cpp line 386 to english please?
Yes, now it's at 396 - 398

fengyvoid committed:
Add EBPMT for PMT pairing in EventBuilding

marc1uk commented:
`RWMRawWaveforms` and `BRFRawWaveforms` (lines 29,30) are both immediately leaked by lines 46,47. 
It would be sensible to either check the return of the corresponding `Get` calls. It may also be worth checking the corresponding pointers are not null before de-referencing them in the following log lines.

Is it worth moving the check on line 76 to line 44; is this just doing needless work if those checks are both false?

i think line 122 should be enclosed with line 121 in `{ }` for line 120 to prevent spurious printouts. 

It seems like it may not be intended that the printout in line 125 may result in the insertion of an element into the AlmostCompletedWavefroms map via `operator[]`.
Same for line 137. Note that `.at` calls will throw in the event that the element doesn't exist - there may be more logic required to avoid that. I'm getting a bit lost with the different potential if conditions. May be better to check & access the map just once and re-use the result.

Would suggest using references on lines 152, 154, 284, 290, 295, 297, 446, 447, 450-453 to avoid copies.

Not sure if the logic prohibits it but would suggest for good practice the use of `emplace` rather than `operator[]` and checking the return value second to avoid potentially clobbering previous entries on lines 325, 326, 327.

fengyvoid committed:
Modifications based on comments from PR#332:

1. I move the get and check of `RWMRawWaveforms` and `BRFRawWaveforms` to line 75 to 85.
2. I move the check and return to line 44-49.
3. The line 122 was still at 122, but only print after the `[]`
4. For line 125 and 137, I change the usage of `[]` to a int AlmostCompleteWaveforms_CountHere, initialized to be zero.
5. I would want to use that but may want to leave this to future optimization... since everything is working right now. (not perfectly because I do see zero PMT waveform from some timestamps, but they can be removed in later steps no matter where they are from)
6. Now at line 339 to 341, changed to emplace as you suggest.

fengyvoid commented:
Appreciate for all the very useful comments and suggestions! I definitely need to learn more about pointers...

1. I move the get and check of `RWMRawWaveforms` and `BRFRawWaveforms` to line 75 to 85.
2. I move the check and return to line 44-49.
3. The line 122 was still at 122, but only print after the `[]`
4. For line 125 and 137, I change the usage of `[]` to a int AlmostCompleteWaveforms_CountHere, initialized to be zero.
5. I would want to use that but may want to leave this to future optimization... since everything is working right now. (not perfectly because I do see zero PMT waveform from some timestamps, but they can be removed in later steps no matter where they are from)
6. Now at line 339 to 341, changed to emplace as you suggest.

fengyvoid committed:
Fix a bug about NumWavesInCompleteSet.
Previously, if NumWavesInCompleteSet was set to 140 and there are only ~133 PMT waveforms exist in data, the AlmostCompleteWaveforms[PMTCounterTimeNs] won't be created. But the usage of AlmostCompleteWaveforms[PMTCounterTimeNs] in Log will create an entry there. This won't affect the later code because there are checks to require AlmostCompleteWaveforms[PMTCounterTimeNs] > 5.

Now the Log was removed and the tolerance for PMT number is changed to 10 at:
ChannelKey.size() >= (NumWavesInCompleteSet - 10)

fengyvoid committed:
The correct file for previous commit

S81D commented:
It looks like Marcus' points were addressed in the new changes. As this is a standalone tool and is in the event building workflow currently in use, I say it's good to go.

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+
+
+
+  if (exeNum % 80 == 0 && exeNum != 0)
```
this is a bit confusing and seems a bit hand wavey. I don't really follow the comment, but am not a great fan of "do the VME offset correction once every X number of executes because that seems to be about right probably".
Is there not a more definitive way of determining when it needs to be done?

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+  vector<uint64_t> RWMEmplacedTimes;
+  vector<uint64_t> BRFEmplacedTimes;
+
+  for (std::pair<uint64_t, std::map<unsigned long, std::vector<Hit>> *> p : *InProgressHits)
```
would prefer using a reference in range-based for loops, but since value is a pointer i suppose this isn't too critical.

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+  {
+    uint64_t PMTCounterTimeNs = p.first;
+    std::map<unsigned long, std::vector<Hit>> *hitMap = p.second;
+    vector<unsigned long> ChannelKey = InProgressChkey->at(PMTCounterTimeNs);
```
again making copies of vectors isn't great efficiency, using a reference would be better.

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+    if (ChannelKey.size() >= (NumWavesInCompleteSet - 10))
+    {
+      Log("EBPMT: ChannelKey.size() == (NumWavesInCompleteSet - 1), ChannelKey.size() = " + std::to_string(ChannelKey.size()) + " == NumWavesInCompleteSet - 1 = " + std::to_string(NumWavesInCompleteSet - 1), v_debug, verbosityEBPMT);
+      if (AlmostCompleteWaveforms.find(PMTCounterTimeNs) != AlmostCompleteWaveforms.end())
```
AlmostCompleteWaveforms is a `std::map` - no need to use `.find() != .end()`, just use `.count()`

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+//      cout<<ch<<", ";
+//    }
+//    cout<<endl;
+auto it_acw = AlmostCompleteWaveforms.find(PMTCounterTimeNs);
```
again shouldn't be using `find` with `std::map`

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+
+      // check if the PMTCounterTimeNs is already in the FinishedHits, if not then add it
+
+      if (FinishedHits->find(PMTCounterTimeNs) != FinishedHits->end())
```
again `std::map.count`

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+      if (PairedPMTTimeStamps.size() > 0)
+      {
+        bool skip = false;
+        for (std::pair<int, vector<uint64_t>> p : PairedPMTTimeStamps)
```
again use reference in range-based for to avoid copy, especially as key in this case is not a pointer

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+        bool skip = false;
+        for (std::pair<int, vector<uint64_t>> p : PairedPMTTimeStamps)
+        {
+          vector<uint64_t> PMTTimeStamps = p.second;
```
another vector copy, could be reference

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+  // loop over all the Hits, for each FinishedHits, loop all grouped triggers, if the time differencs<100 for trigger 5,
+  // matchedHitNumber ++, then remove the hit from the map
+  int loopNum = 0;
+  for (std::pair<uint64_t, std::map<unsigned long, std::vector<Hit>> *> pmtpair : *FinishedHits)
```
use reference - not critical as key is pointer

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+    int matchedTrack = 0;
+    int matchedIndex = 0;
+    // loop all tracks of GroupedTriggersInTotal
+    for (std::pair<int, std::vector<std::map<uint64_t, uint32_t>>> pair : GroupedTriggersInTotal)
```
use reference to avoid copy of vector of map

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+      if (TrackTriggerWord != matchToTrack && !matchToAllTriggers)
+        continue;
+
+      vector<std::map<uint64_t, uint32_t>> groupedTriggers = pair.second;
```
use reference

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+      // loop all the grouped triggers, if the value is target trigger, then calculate the time difference
+      for (int i = 0; i < groupedTriggers.size(); i++)
+      {
+        map<uint64_t, uint32_t> groupedTrigger = groupedTriggers.at(i);
```
use reference

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+    if (saveRWMWaveforms)
+    {
+      // check does RWMRawWaveforms have key PMTCounterTimeNs, if not, print a log waring and skip
+      if (RWMRawWaveforms->find(PMTCounterTimeNs) == RWMRawWaveforms->end())
```
`std::map.count()` instead of `.find()`

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+    if (saveBRFWaveforms)
+    {
+      // check does BRFRawWaveforms have key PMTCounterTimeNs, if not, print a log waring and skip
+      if (BRFRawWaveforms->find(PMTCounterTimeNs) == BRFRawWaveforms->end())
```
count instead of find

marc1uk commented on UserTools/EBPMT/EBPMT.cpp:
```diff
@@ -0,0 +1,513 @@
+
+  Log("EBPMT: Found " + std::to_string(timestamps.size()) + " timestamps", v_message, verbosityEBPMT);
+
+  // loop timestamps，对于每一个时间戳，检查它与它之前的时间戳的差值是否是8或者16
```
english translation please

marc1uk commented on UserTools/Unity.h:
```diff
@@ -182,3 +182,5 @@
 #include "AssignBunchTimingMC.h"
 #include "FitRWMWaveform.h"
+#include "EBPMT.h"
+#include "EBPMT.h"
```
duplicate entry

marc1uk committed:
Merge branch 'Application' into EBPMT_PR

marc1uk commented:
I'm inclined to merge this to avoid holding things off any longer. The requested changes are very minimal - both with regards to time/effort to implement, but also in terms of severity. In light of this, I'm going to merge, but as per #333 I would appreciate if someone can make (and validate) the requested changes in a new PR.
